### PR TITLE
Spanner instance direct controller

### PR DESCRIPTION
**Pull request by anhdle-sso**
### Change description

Fixes #

### Tests you have done

- [x] Run `make ready-pr` to ensure this PR is ready for review.
- [x] Perform necessary E2E testing for changed resources.

### CODE REVIEW

File: apis/spanner/v1beta1/instance_reference.go [hunk 1]
```diff
@@ -168,6 +168,10 @@ func parseSpannerInstanceExternal(external string) (parent *SpannerInstanceParen
 	return parent, resourceID, nil
 }
 
+func (r *SpannerInstanceRef) SpannerInstanceConfigPrefix() string {
+	return fmt.Sprintf("projects/%s/instanceConfigs/", r.parent.ProjectID)
```
Comment (yuwenma) [inline apis/spanner/v1beta1/instance_reference.go:172]: hmmm, is this the same as `r.parent`?

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_spannerinstances.spanner.cnrm.cloud.google.com.yaml [hunk 2]
```diff
@@ -77,10 +77,10 @@ spec:
                   in length.
                 type: string
               numNodes:
-                format: int64
+                format: int32
                 type: integer
               processingUnits:
-                format: int64
+                format: int32
```
Comment (yuwenma) [inline config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_spannerinstances.spanner.cnrm.cloud.google.com.yaml:83]: I don't remember the details, could you verify against real k8s cluster to see if we can change this? 
e.g. 
1. `kubectl apply crds/<old-spanner-int64>` 
2. `kubectl apply -f <example.yaml using processingUnits int64>` 
3. `kubectl apply crds/<new-spanner-int32>`

File: hack/record-gcp [hunk 3]
```diff
@@ -27,7 +27,7 @@ echo "TEST_BILLING_ACCOUNT_ID=${TEST_BILLING_ACCOUNT_ID}"
 export TEST_BILLING_ACCOUNT_ID
 
 if [[ -z "${KCC_USE_DIRECT_RECONCILERS:-}" ]]; then
-  KCC_USE_DIRECT_RECONCILERS="ComputeForwardingRule"
+  KCC_USE_DIRECT_RECONCILERS="ComputeForwardingRule,SpannerInstance"
```
Comment (yuwenma) [inline hack/record-gcp:30]: Suggest not changing this env var.  We don't run real gcp against presubmit, so this is more like a local setup.

File: pkg/test/resourcefixture/testdata/basic/spanner/v1beta1/spannerinstance/_http.log [hunk 4]
```diff
@@ -164,28 +170,29 @@ X-Xss-Protection: 0
...
     "labels": {
       "cnrm-test": "true",
       "label-one": "value-one",
       "managed-by-cnrm": "true"
     },
-    "name": "projects/${projectId}/instances/%!s(\u003cnil\u003e)",
-    "nodeCount": 3,
-    "processingUnits": 2000
+    "name": "projects/${projectId}/instances/spannerinstance-sample-${uniqueId}",
+    "nodeCount": 3
```
Comment (yuwenma) [inline pkg/test/resourcefixture/testdata/basic/spanner/v1beta1/spannerinstance/_http.log:195]: nice

File: apis/spanner/v1beta1/instance_types.go
Comment (yuwenma) [inline apis/spanner/v1beta1/instance_types.go:1]: Can you add a "NOTYET" comment here about adding `status.observedState.processingUnites` and `status.observedState.numNodes` in a follow up PR? I think it is essential to have those two fields considering the logic around these two fields (and the later on autoscaling)

File: pkg/controller/direct/spanner/instance_controller.go [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+	if err != nil {
+		return nil, err
+	}
+	resourceID := direct.ValueOf(obj.Spec.ResourceID)
+	if resourceID == "" {
+		resourceID = obj.GetName()
+	}
+	if resourceID == "" {
+		return nil, fmt.Errorf("cannot resolve resource ID")
+	}
```
Comment (yuwenma) [inline pkg/controller/direct/spanner/instance_controller.go:89]: This should be from the NewSpannerInstanceRef

File: pkg/controller/direct/spanner/instance_controller.go [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+	desired := a.desired.DeepCopy()
+	if err := a.SpecValidation(); err != nil {
+		return err
+	}
+	resource := SpannerInstanceSpec_ToProto(mapCtx, &desired.Spec, a.id.SpannerInstanceConfigPrefix())
+	// If node count or processing unit is not specify,
+	// Default NodeCount to 1.
+	if resource.NodeCount == 0 && resource.ProcessingUnits == 0 {
+		resource.NodeCount = 1
+	}
```
Comment (yuwenma) [inline pkg/controller/direct/spanner/instance_controller.go:144]: IIRC, the GCP service sets the default nodeCount=1, processingUnites 1000. If so, KCC does not need to duplicate this default.

File: pkg/controller/direct/spanner/instance_controller.go [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+	resource := SpannerInstanceSpec_ToProto(mapCtx, &desired.Spec, a.id.SpannerInstanceConfigPrefix())
+	resource.Name = a.id.External
+	resource.Labels = desired.Labels
+	resource.Labels["managed-by-cnrm"] = "true"
+	resource.Name = a.id.External
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	updateMask := &fieldmaskpb.FieldMask{}
```
Comment (yuwenma) [inline pkg/controller/direct/spanner/instance_controller.go:198]: Suggest using the `common.CompareProtoMessage` here?

File: apis/spanner/v1beta1/instance_types.go
Comment (anhdle-sso) [reply apis/spanner/v1beta1/instance_types.go:1]: ack

File: apis/spanner/v1beta1/instance_reference.go (hunk 1 above)
Comment (anhdle-sso) [reply apis/spanner/v1beta1/instance_reference.go:172]: not actually, the parent of instance is just projects/<project>. The instance config is a different pseudo resource. The Create function need the config field in this form to work.

File: pkg/controller/direct/spanner/instance_controller.go (hunk 7 above)
Comment (yuwenma) [reply pkg/controller/direct/spanner/instance_controller.go:195]: BTW, we can improve the CompareProtoMessage to apply comparison with customized conditions.

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_spannerinstances.spanner.cnrm.cloud.google.com.yaml (hunk 2 above)
Comment (yuwenma) [reply config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_spannerinstances.spanner.cnrm.cloud.google.com.yaml:83]: That's awesome, thank you!

**Review by yuwenma — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**